File: backend/src/main/kotlin/br/com/stocksense/client/dto/PredictResponse.kt
package br.com.stocksense.client.dto

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.databind.annotation.JsonNaming
import java.math.BigDecimal
import java.time.LocalDate

/**
 * Contrato de saída do `POST /predict` do ml-service. Espelha o Pydantic
 * `PredictResponse`.
 *
 * `@JsonIgnoreProperties(ignoreUnknown = true)`: mantido por tolerância a campos
 * futuros — o ml-service removeu `classe_abc`/`abc_proxy` do response (a dívida
 * técnica da ADR #3 foi resolvida do lado dele; nunca existiram neste DTO).
 * `desvio_padrao_demanda` (T-05) passou a ser devolvido pelo ml-service e agora
 * é mapeado abaixo.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
data class PredictResponse(
    val produtoId: Int,
    val modeloSelecionado: String,
    val previsoes: List<PrevisaoDiaria>,
    val metricas: Map<String, MetricasModelo>,
    val pontoReposicao: BigDecimal,
    val estoqueSeguranca: BigDecimal,
    val diasAteRuptura: BigDecimal? = null,
    val desvioPadraoDemanda: BigDecimal? = null,
    val aviso: String? = null,
)

@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
data class PrevisaoDiaria(
    val data: LocalDate,
    val quantidadePrevista: BigDecimal,
)

data class MetricasModelo(
    val mape: BigDecimal,
    val rmse: BigDecimal,
    val mae: BigDecimal,
)

File: ml-service/app/models/predict_response.py
from datetime import date

from pydantic import BaseModel


class MetricasModelo(BaseModel):
    """Métricas de avaliação de um modelo de séries temporais."""

    mape: float
    rmse: float
    mae: float


class PrevisaoDiaria(BaseModel):
    """Quantidade prevista para um único dia."""

    data: date
    quantidade_prevista: float


class PredictResponse(BaseModel):
    """Payload de saída do endpoint POST /predict."""

    produto_id: int
    modelo_selecionado: str
    previsoes: list[PrevisaoDiaria]
    metricas: dict[str, MetricasModelo]
    ponto_reposicao: float
    estoque_seguranca: float
    dias_ate_ruptura: float | None
    desvio_padrao_demanda: float
    aviso: str | None = None
